### PR TITLE
Integration test with chain simulator

### PR DESCRIPTION
**Pull request by sstanculeanu**

### CODE REVIEW

File: integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go [hunk 1]
```diff
@@ -0,0 +1,798 @@
...
+
+	// deploy multisig
+	minRelayerStakeInt, _ := big.NewInt(0).SetString(minRelayerStake, 10)
+	minRelayerStakeHex := hex.EncodeToString(minRelayerStakeInt.Bytes())
+	multisigAddress, err := multiversXProxyWithChainSimulator.DeploySC(
+		ctx,
+		multisigContract,
+		ownerKeys.pk,
+		ownerKeys.sk,
+		[]string{wasm.VMTypeHex, "0500", getHexAddress(t, safeAddress), getHexAddress(t, multiTransferAddress), minRelayerStakeHex, slashAmount, quorum, getHexAddress(t, relayersKeys[0].pk), getHexAddress(t, relayersKeys[1].pk), getHexAddress(t, relayersKeys[2].pk)},
```
Comment (iulianpascalau) [inline integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go:384]: small for loop for the relayers public keys?

File: integrationTests/relayers/testdata/config/nodeConfig/upgradeContracts/dns/v3.0/deploy.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (iulianpascalau) [inline integrationTests/relayers/testdata/config/nodeConfig/upgradeContracts/dns/v3.0/deploy.json:1]: can be removed along with anything on upgradeContracts directory

File: integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+	ownerPem                     = "testdata/wallets/owner.pem"
+	safeContract                 = "testdata/contracts/esdt-safe.wasm"
+	multisigContract             = "testdata/contracts/multisig.wasm"
+	multiTransferContract        = "testdata/contracts/multi-transfer-esdt.wasm"
+	bridgeProxyContract          = "testdata/contracts/bridge-proxy.wasm"
+	aggregatorContract           = "testdata/contracts/aggregator.wasm"
+	wrapperContract              = "testdata/contracts/bridged-tokens-wrapper.wasm"
+	nodeConfig                   = "testdata/config/nodeConfig"
+	proxyConfig                  = "testdata/config/proxyConfig"
+	minRelayerStake              = "10000000000000000000" // 10egld
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go]: done

File: integrationTests/proxyWithChainSimulator.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+	blocksGenerator        process.BlocksGenerator
+	proxyInstance          multiversx.Proxy
+	simulator              chainSimulatorHandler
+	pkConv                 core.PubkeyConverter
+	roundDuration          time.Duration
+}
+
+// CreateProxyWithChainSimulator creates a new instance of chain simulator with proxy
+func CreateProxyWithChainSimulator(args ArgProxyWithChainSimulator) (*proxyWithChainSimulator, error) {
+	roundDurationInMillis := args.RoundDurationInMs
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/proxyWithChainSimulator.go:68]: yes, so we don't have to handle the blocks generation

File: integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+	relayers := startRelayers(t, numRelayers, multiversXProxyWithChainSimulator, ethereumChainMock, safeContractEthAddress, erc20ContractsHolder, safeAddress, multisigAddress)
+	defer closeRelayers(relayers)
+
+	// wait for signal interrupt or time out
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, syscall.SIGINT, syscall.SIGTERM)
+	for {
+		select {
+		case <-interrupt:
+			log.Error("signal interrupted")
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go:206]: added

File: integrationTests/proxyWithChainSimulator.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+	emptyAddress, err := instance.pkConv.Encode(make([]byte, 32))
+	if err != nil {
+		return "", err
+	}
+
+	scCode := wasm.GetSCCode(path)
+	params := []string{scCode}
+	params = append(params, parameters...)
+	txData := strings.Join(params, "@")
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/proxyWithChainSimulator.go:198]: renamed

File: integrationTests/proxyWithChainSimulator.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+	if errGet != nil {
+		return "", errGet
+	}
+
+	return txResult.Logs.Events[0].Address, nil
+}
+
+// GetTransactionResult tries to get a transaction result. It may wait a few blocks
+func (instance *proxyWithChainSimulator) GetTransactionResult(hash string) (*transaction.ApiTransactionResult, error) {
+	txResult, errGet := instance.simulator.GetNodeHandler(0).GetFacadeHandler().GetTransaction(hash, true)
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/proxyWithChainSimulator.go:229]: forgot about it, updated as suggested

File: integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go [hunk 8]
```diff
@@ -0,0 +1,798 @@
...
+
+func executeContractsTxs(
+	t *testing.T,
+	ctx context.Context,
+	multiversXProxyWithChainSimulator proxyWithChainSimulator,
+	relayersKeys []keysHolder,
+	ownerKeys keysHolder,
+) (string, string, string, string) {
+	// fund the involved wallets(owner + relayers)
+	multiversXProxyWithChainSimulator.FundWallets([]string{
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go:314]: updated

File: integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go [hunk 9]
```diff
@@ -0,0 +1,798 @@
...
+	// wait for epoch 1 before sc deploys
+	time.Sleep(time.Duration(roundDurationInMs*(roundsPerEpoch+2)) * time.Millisecond)
+
+	// deploy aggregator
+	aggregatorAddress, err := multiversXProxyWithChainSimulator.DeploySC(
+		ctx,
+		aggregatorContract,
+		ownerKeys.pk,
+		ownerKeys.sk,
+		[]string{wasm.VMTypeHex, "0500", "01", "00", getHexAddress(t, ownerKeys.pk)},
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/relayers/ethToMultiversXWithChainSimulator_test.go:330]: updated

File: integrationTests/proxyWithChainSimulator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply integrationTests/proxyWithChainSimulator.go:68]: great :+1:

**Review by iulianpascalau — DISMISSED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**